File: ddd/tests/fixtures/operation-error-set/typescript-companion-workspace/billing-domain/src/contract.ts
/**
 * The command declares no result contract, and the generation method returns
 * its own type instead of the standard result.
 */

export type IssueInvoiceError = "already-issued" | "empty-lines";
export type OpenInvoiceError = "negative-amount" | "missing-customer";

const invoiceBrand: unique symbol = Symbol("Invoice");

export type Invoice = {
  readonly [invoiceBrand]: true;
  issue(): void;
};

export const Invoice = {
  open(): Invoice {
    return {
      [invoiceBrand]: true,
      issue() {},
    };
  },
};

File: ddd/tests/fixtures/operation-error-set/rust-workspace/billing-domain/src/contract.rs
//! The command declares no result contract, and the generation method returns
//! its own type instead of the standard result.

pub enum IssueInvoiceError {
    AlreadyIssued,
    EmptyLines,
}

pub enum OpenInvoiceError {
    NegativeAmount,
    MissingCustomer,
}

pub struct Invoice;

impl Invoice {
    pub fn issue(&mut self) {}

    pub fn open() -> Self {
        Self
    }
}
